src/cmd/server/start.ts

```diff
@@ -25,41 +29,96 @@ export default async (
+  // Source: https://github.com/moxystudio/node-proper-lockfile#user-content-comparison
+  let fd: number|null
+  try {
+    fd = await open(serverJsonPath, 'wx')
+  } catch (error) {
+    if (error.code !== 'EEXIST') {
+      throw error
+    }
+    // Another process won the race, stop this process.
+    return
   }
-  const serverJsonPath = path.join(connectionInfoDir, 'server.json')
-  await writeJsonFile(serverJsonPath, {
-    connectionOptions,
-    pid: process.pid,
-  })
+  try {
+    let server: null|{close (): Promise<void>} = null
+    onExit(() => {
+      if (server !== null) {
+        // Note that server.close returns a Promise, but we cannot wait for it because we may be
+        // inside the 'exit' even of process.
+        server.close()
+      }
+      if (fd !== null) {
+        const fdForClose = fd
+        // Set fd to null so we only attempt to close it once.
+        fd = null
+        try {
+          fs.closeSync(fdForClose)
+        } catch (error) {
+          logger.error(error, `Got error while closing file descriptor ${fdForClose}, but the process is already exiting`)
+        }
+      }
+      fs.unlinkSync(serverJsonPath)
+    })
+    const store = await createStore(Object.assign(opts, {
+      store: pathOfStore,
+    }))
+    const protocol = opts.protocol || opts.port && 'tcp' || 'auto'
 
-  const server = createServer(store.ctrl, {
-    ...serverOptions,
-    ignoreStopRequests: opts.ignoreStopRequests,
-    ignoreUploadRequests: opts.ignoreUploadRequests,
-  })
+    // If the protocol is 'tcp' and opts.port is not specified then a random port is chosen.
+    // Choosing a random port is done by starting and stopping a server. But we are already starting
+    // a server, so effectively we are doing start, stop, start. We should remove the 'get-port' NPM
+    // module, write server.json after createServer and return the TCP port from createServer.
+    const serverOptions = await getServerOptions(connectionInfoDir, {protocol, port: opts.port})
+    const connectionOptions = {
+      remotePrefix: serverOptions.path
+        ? `http://unix:${serverOptions.path}:`
+        : `http://${serverOptions.hostname}:${serverOptions.port}`,
+    }
+    server = createServer(store.ctrl, {
+      ...serverOptions,
+      ignoreStopRequests: opts.ignoreStopRequests,
+      ignoreUploadRequests: opts.ignoreUploadRequests,
+    })
+    // Make sure to populate server.json after the server has started, so clients know that the server is
+    // listening if a server.json with valid JSON content exists.
+    await writeFile(fd,
+      JSON.stringify({
+        connectionOptions,
+        pid: process.pid,
+      }, undefined, 2), // undefined and 2 are for formatting.
+      'utf8')
+    await tryCloseFd(false)
+  } catch (error) {
+    await tryCloseFd(true)
+    throw error
+  }
 
-  onExit(() => {
-    server.close()
-    fs.unlinkSync(serverJsonPath)
-  })
+  async function tryCloseFd (handleErrors: boolean) {
```

Comment:
instead of the handleErrors boolean, just create a new function that closes the file and call it from the "ignore errors" close function. You have to functions. closeServerJson() and silentCloseServerJson()